ansibulled/cli/ansibulled_changelog.py

```diff
@@ -205,6 +206,8 @@ def command_release(args: Any) -> None:
             galaxy = load_galaxy_metadata(paths)
             version = galaxy['version']
 
+    version = cast(str, version)
+
```

Comment:
I'm not a big fan of setting type far away from where the variable is set and from the documentation of typing.cast() I'm also not sure if it's a good idea to use it except as an escape hatch when the code is really problematic otherwise.  Here's a few thoughts:

* I think setting the type of a variable belongs in the same place as the assignment to the variable otherwise the type and variable setting can get out of sync.
  * For instance, let's say that in the future we remove the `elif not version`.  if the cast is here, outside of that block, then the type checker won't realize that version could now contain None.
* The documentation of cast says that it is unconditionally telling the type checker to treat this value as the new value instead of the value it actually has.  This seems wrong..
  * For instance, if new code is added above which explicitly sets version to None.  This cast will hide that fact by telling the type checker to treat it as if it's a str instead.
* I think type checking in python should be about contracts.  If we do `version, codename = get_ansible_release()` what we want to say is "enforce a contract that the return from get_ansible_release will only put a str into version."  What cast is actually doing the opposite, it's saying "whatever value is in version, treat it as if it's a str."

This is where I think what I had written about the difference between mypy and pyre comes into play. I think that the code should look like this:

```python
    if not version or not codename:
        if not config.is_collection:
            # Both version and codename are required for Ansible (Base)
            try:
                version: str, codename = get_ansible_release()
            except ValueError:
                LOGGER.error('Cannot import ansible.release to determine version and codename')
                sys.exit(3)
        elif not version:
            # Codename is not required for collections, only version is
            galaxy = load_galaxy_metadata(paths)
            version: str = galaxy['version']

```

In the two places where we rebind version to a new value, we set up a new contract that says that the new value must now be a str, rather than either a str or None.  In addition to avoiding the problems that case had, it also means that if get_ansible_release() returns None, a type error will be flagged against line 199 instead of much further away, where the None is used.

Reply:
`version: str = galaxy['version']` does not work for mypy, because the variable has already been defined (and types have to be specified at point of definition, i.e. when first set). What does work would be `version = cast(str, galaxy['version'])`. Instead, I rewrote that code anyway and added proper checking. (Also, the `load_galaxy_metadata` could result in an exception which needs to be caught.)

Reply:
> `version: str = galaxy['version']` does not work for mypy, because the variable has already been defined

Right. which is why I'm against using myoy :-). It is unable to handle the case of refining the contract for a variable and so it causes you to rewrite your code badly in order to accommodate it.

This might be a missing feature for mypy or it might be that myoy actually has a vision of types that doesn't match with python's architecture, I'm not sure.  I only got far enough to find that pyre does this right while mypy does it wrong.

Good work getting rid of cast, though. I think even in your example of how to write cast in a way mypy likes, it would have been doing the wrong thing: it would not update the contact, but instead have told the type checker to lie about the type of the data.  Is there import of cast at the top of the file still needed?

Reply:
Ah, i see some existing coffee uses cast... That code looks wrong too. paths.galaxy_path probably needs to be guaranteed to be a str by that time.  (Or there needs to be an additional conditional to remove non-str values)

Reply:
The problem is that the contract cannot be modelled with type hinting except by rewriting part of the code. `paths.galaxy_path` is currently not `None` iff `paths.is_collection` is `True`.

Anyway, I want to adjust the program to not assume that `galaxy.yml` is around for collections, since [there are people who want to template that file only when needed](https://github.com/ansible-collections/overview/pull/56). Something that the changelog tool right now does not support.

Reply:
IIUC, I think what you're describing is a bug in the code, not a bug in the type hints.  The type hints are telling us that paths.galaxy_path could be None and when that happens, it would cause `os.path.exists` to fail and traceback.  So the type checker is right to fail the code as is.

Reply:
That's not (really) true. The type checker simply has insufficient information, since it is not possible to model certain contracts ("attribute X is not `None` when attribute Y has value Z"). Such contracts are notoriously hard to represent without rewriting code (and leads to bugs when the contract is not satisfied somewhere by accident).

Reply:
For example, the CheckerFramework for Java allows to model some simple contracts like "expression X is not `null` when function Y returned value Z": https://checkerframework.org/api/org/checkerframework/checker/nullness/qual/EnsuresNonNullIf.html
This is not possible with the typing hints in Python.

Reply:
I agree that in general there's contracts which can be hard to model but I think we can model the specific contract in`command_init()` without cast...  When I read the code, I see something like this pseudo-code:
``` 
if not args.root or \
  (changelog_dir in base_dir and config_dir in base_dir and 'lib/ansible' not in base_dir):
    paths.galaxy = None
paths.galaxy = 'some string'
```
That's the code path from `command_init()` => `set_paths()` => `PathsConfig`[`force_collection()` or `detect()`]

So... the type checker would be correct to throw an error on line 142 because `galaxy_path` could be None and `os.path.exists()` would error if that were the case.  (Line 142 is):
``` python
     if not os.path.exists(cast(str, paths.galaxy_path)):
```

The `cast` in Line 142 is hiding the legitimate error.  When I look at what the test is doing, it seems like there is an easy solution which satisfies the type checker:

``` python
if paths.galaxy_path is None or not os.path.exists(paths.galaxy_path):
    LOGGER.error('The file galaxy.yml does not exists in the collection root!')
    sys.exit(3)
```

Since this check needs galaxypath to exist, otherwise it exits, if the galaxy_path is None, then we can also tell that it doesn't exist.  The type checker (both pyre and mypy) infer a new contract for galaxy_path when they see the simple condition which filters out whether the path is None. ( https://mypy.readthedocs.io/en/stable/kinds_of_types.html#optional-types-and-the-none-type )

Do you think there's something I'm still missing?